euphonic/spectra.py

```diff
@@ -669,10 +674,387 @@ def broaden(self: T, x_width,
+        init_kwargs = {
+            self._spectrum_data_name(): self._get_spectrum_data()[item, :],
+            "x_tick_labels": self.x_tick_labels,
+            "metadata": self._get_item_metadata(item)
+                       } | self._get_bin_kwargs()
+
+        if isinstance(item, Integral):
+            return self._item_type(**init_kwargs)
+
+        return type(self)(**init_kwargs)
+
+    def _validate_item(self, item: Integral | slice | Sequence[Integral] | np.ndarray
+                       ) -> None:
+        """Raise Error if index has inappropriate typing/range"""
+        if isinstance(item, Integral):
+            return
+        if isinstance(item, slice):
+            if (item.stop is not None) and (item.stop >= len(self)):
+                raise IndexError(f'index "{item.stop}" out of range')
+            return
+
+        if not all(isinstance(i, Integral) for i in item):
+            raise TypeError(
+                f'Index "{item}" should be an integer, slice '
+                f'or sequence of ints')
+
+    @overload
+    def _get_item_metadata(self, item: Integral) -> OneLineData:
+        """Get a single metadata item with no line_data"""
+
+    @overload
+    def _get_item_metadata(self, item: slice | Sequence[Integral] | np.ndarray
+                           ) -> Metadata:  # noqa: F811
+        """Get a metadata collection (may include line_data)"""
+
+    def _get_item_metadata(self, item):  # noqa: F811
+        """Produce appropriate metadata for __getitem__"""
+        metadata_lines = list(self.iter_metadata())
+
+        if isinstance(item, Integral):
+            return metadata_lines[item]
+        if isinstance(item, slice):
+            return self._combine_metadata(metadata_lines[item])
+        if len(item) == 1:
+            return metadata_lines[item[0]]
+        return self._combine_metadata(
+            list(itemgetter(*item)(metadata_lines)))
+
+    def copy(self) -> Self:
+        """Get an independent copy of spectrum"""
+        return self._item_type.copy(self)
+
+    def __add__(self, other: Self) -> Self:
+        """
+        Appends the y_data of 2 Spectrum1DCollection objects,
+        creating a single Spectrum1DCollection that contains
+        the spectra from both objects. The two objects must
+        have equal x_data axes, and their y_data must
+        have compatible units and the same number of y_data
+        entries
+
+        Any metadata key/value pairs that are common to both
+        spectra are retained in the top level dictionary, any
+        others are put in the individual 'line_data' entries
+        """
+        return type(self).from_spectra([*self, *other])
+
+    def iter_metadata(self) -> Generator[OneLineData, None, None]:
+        """Iterate over metadata dicts of individual spectra from collection"""
+        common_metadata = {
+            key: self.metadata[key]
+            for key in set(self.metadata.keys()) - {"line_data",}}
```

Comment:
Looking at this again with not midnight eyes.
```suggestion
            for key in self.metadata
            if key != "line_data"
            }
```

Reply:
🤦 

Ok, done but with two fewer linebreaks.

Reply:
Here's another option with less repetition:
```python
        from toolz.dicttoolz import keyfilter
        common_metadata = keyfilter(lambda key: key != "line_data",
                                    self.metadata)
```

Reply:
Here, I think you're accidentally replacing standard Python with toolz for the sake of it. 

I'm all for functional as you know, but I think here this is obfuscated to people who don't know toolz adding an extra setop, the more standard solution is better if only for familiarity. 

Plus a lambda and a function calls rather than a simple comprehension isn't reducing complexity for me.

Reply:
Before I revert that, take a look at the new version of `_combine_metadata` I just pushed. In this case the `keyfilter` avoids nested comprehensions so I think the clarity gain is better.

But still perhaps the overall difference is too marginal to be worth it.

Reply:
Getting a bit silly, but maybe this is part of the process of finding a good version:

```python
        # Put all other per-spectrum metadata in line_data
        from toolz.curried import map
        from toolz.functoolz import pipe

        keep_uncommon_metadata_items = keyfilter(
            lambda key: key not in common_metadata)

        line_data = pipe(all_metadata,
                         map(keep_uncommon_metadata_items),
                         list)

```

Reply:
Taking away the Extremely Functional parts but keeping the named keyfilter function:

```python
        # Put all other per-spectrum metadata in line_data
        keep_uncommon_metadata_items = keyfilter(
            lambda key: key not in common_metadata)

        line_data = list(map(keep_uncommon_metadata_items, all_metadata))
```

Reply:
I still think we're better off with:
```python
common_metadata = {
            key: self.metadata[key] for key in self.metadata if key != "line_data"
}
```